lib/models/database/user_repository.dart

```diff
@@ -0,0 +1,147 @@
+
+  const UserFirestore({required this.uid, required this.data});
+
+  factory UserFirestore.fromDb(DocumentSnapshot docSnap) {
+    if (!docSnap.exists) {
+      throw Exception("User document does not exist");
+    }
+
+    try {
+      final data = docSnap.data() as Map<String, dynamic>;
+
+      return UserFirestore(
+        uid: docSnap.id,
+        data: UserFirestoreData.fromDbData(data),
+      );
+    } catch (e) {
+      if (e is TypeError) {
+        throw Exception("Cannot parse user document : ${e.toString()}");
+      } else {
+        rethrow;
+      }
+    }
+  }
+
+  @override
+  bool operator ==(Object other) {
+    return other is UserFirestore && other.uid == uid && other.data == data;
+  }
+
+  @override
+  int get hashCode => Object.hash(uid, data);
+}
+
+@immutable
+class UserFirestoreData {
+  final String username;
+  static const String usernameField = "username";
+
+  final String displayName;
+  static const String displayNameField = "displayName";
+
+  final Timestamp joinTime;
+  static const joinTimeField = "joinTime";
+
+  const UserFirestoreData({
+    required this.username,
+    required this.displayName,
+    required this.joinTime,
+  });
+
+  factory UserFirestoreData.fromDbData(Map<String, dynamic> data) {
+    try {
+      return UserFirestoreData(
+        username: data[usernameField],
+        displayName: data[displayNameField],
+        joinTime: data[joinTimeField],
+      );
+    } catch (e) {
+      if (e is TypeError) {
+        throw Exception("Cannot parse user data document: ${e.toString()}");
+      } else {
+        rethrow;
+      }
+    }
+  }
+
+  Map<String, dynamic> toDbData() {
+    return {
+      usernameField: username,
+      displayNameField: displayName,
+      joinTimeField: joinTime,
+    };
+  }
+
+  @override
+  bool operator ==(Object other) {
+    return other is UserFirestoreData &&
+        other.username == username &&
+        other.displayName == displayName &&
+        other.joinTime == joinTime;
+  }
+
+  @override
+  int get hashCode => Object.hash(username, displayName, joinTime);
+}
+
+class UserRepository {
+  final FirebaseAuth _firebaseAuth;
+
+  static const collectionName = "users";
+  final CollectionReference _collectionRef;
+
+  UserRepository({
+    required FirebaseFirestore firestore,
+    required FirebaseAuth firebaseAuth,
+  })  : _firebaseAuth = firebaseAuth,
+        _collectionRef = firestore.collection(collectionName);
+
+  Future<UserFirestore> getCurrentUser() async {
+    final uid = _getUid();
+
+    return getUser(uid);
+  }
+
+  Future<UserFirestore> getUser(UserFirestoreId uid) async {
+    final docSnap = await _collectionRef.doc(uid).get();
+
+    return UserFirestore.fromDb(docSnap);
+  }
+
+  Future<void> setCurrentUser(UserFirestoreData userData) async {
+    final uid = _getUid();
+
+    await _collectionRef.doc(uid).set(userData.toDbData());
+  }
+
+  Future<bool> doesUserExists(UserFirestoreId uid) async {
```

Comment:
Small typo, this should be `doesUserExist` (without an s at the end)

Reply:
Yep, corrected in 47a4e5bd8a4449a5739b4af036d8021f70b5ba59 :+1: .